src/extensions/default/NavigationAndHistory/NavigationProvider.js

```diff
@@ -0,0 +1,553 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
+ * DEALINGS IN THE SOFTWARE.
+ * 
+ */
+ 
+/**
+ * Manages Editor navigation history to aid back/fwd movement between the edit positions 
+ * in the active project context. The navigation history is purely in-memory and not 
+ * persisted to file system when a project is being closed.
+ */
+define(function (require, exports, module) {
+    "use strict";
+
+    var Strings                 = brackets.getModule("strings"),
+        MainViewManager         = brackets.getModule("view/MainViewManager"),
+        Document                = brackets.getModule("document/Document"),
+        DocumentManager         = brackets.getModule("document/DocumentManager"),
+        DocumentCommandHandlers = brackets.getModule("document/DocumentCommandHandlers"),
+        EditorManager           = brackets.getModule("editor/EditorManager"),
+        Editor                  = brackets.getModule("editor/Editor"),
+        ProjectManager          = brackets.getModule("project/ProjectManager"),
+        CommandManager          = brackets.getModule("command/CommandManager"),
+        Commands                = brackets.getModule("command/Commands"),
+        Menus                   = brackets.getModule("command/Menus"),
+        KeyBindingManager       = brackets.getModule("command/KeyBindingManager"),
+        FileSystem              = brackets.getModule("filesystem/FileSystem");
+    
+    var KeyboardPrefs = JSON.parse(require("text!keyboard.json"));
+    
+    // Command constants for navigation history
+    var NAVIGATION_JUMP_BACK      = "navigation.jump.back",
+        NAVIGATION_JUMP_FWD       = "navigation.jump.fwd";
+    
+    // The latency time to capture an explicit cursor movement as a navigation frame
+    var NAV_FRAME_CAPTURE_LATENCY = 2000,
+        MAX_NAV_FRAMES_COUNT = 30;
+    
+   /**
+    * Contains list of most recently known cursor positions.
+    * @private
+    * @type {Array.<Object>}
+    */
+    var jumpToPosStack = [];
+    
+   /**
+    * Contains list of most recently traversed cursor positions using NAVIGATION_JUMP_BACK command.
+    * @private
+    * @type {Array.<Object>}
+    */
+    var jumpedPosStack = [],
+        activePosNotSynced = false,
+        captureTimer = null,
+        currentEditPos = null,
+        jumpInProgress = false,
+        commandJumpBack,
+        commandJumpFwd;
+        
+   /**
+    * Function to check if there are any navigatable frame backward.
+    * @private
+    */
+    function _hasNavBackFrames() {
+        return (jumpedPosStack.length > 0 && jumpToPosStack.length > 0)
+            || (!jumpedPosStack.length && jumpToPosStack.length > 1);
```

Comment:
I don't quite get this code.
So, when you can't jump forward, there need to be at least two elements in `jumpToPosStack`? Why is that?